llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -1244,39 +1244,60 @@ translateSpirvGlobalUses(LoadInst *LI, StringRef SpirvGlobalName,
   }
 
   // Only loads from _vector_ SPIRV globals reach here now. Their users are
-  // expected to be ExtractElementInst only, and they are replaced in this loop.
-  // When loads from _scalar_ SPIRV globals are handled here as well, the users
-  // will not be replaced by new instructions, but the GenX call replacing the
-  // original load 'LI' should be inserted before each user.
+  // expected to be ExtractElementInst or TruncInst only, and they are replaced
+  // in this loop. When loads from _scalar_ SPIRV globals are handled here as
+  // well, the users will not be replaced by new instructions, but the GenX call
+  // replacing the original load 'LI' should be inserted before each user.
   for (User *LU : LI->users()) {
-    ExtractElementInst *EEI = cast<ExtractElementInst>(LU);
+    assert(
+        (isa<ExtractElementInst>(LU) || isa<TruncInst>(LU)) &&
+        "SPIRV global users should be either ExtractElementInst or TruncInst");
+    Instruction *EEI = cast<Instruction>(LU);
     NewInst = nullptr;
 
+    uint64_t IndexValue = 0;
+    if (isa<ExtractElementInst>(EEI)) {
+      IndexValue = getIndexFromExtract(cast<ExtractElementInst>(EEI));
+    } else {
+      auto *GEPCE =
+          dyn_cast<GetElementPtrConstantExpr>(LI->getPointerOperand());
+      if (GEPCE) {
+        IndexValue = cast<Constant>(GEPCE->getOperand(2))
+                         ->getUniqueInteger()
+                         .getZExtValue();
```

Comment:
The above code silently produces invalid IR if it gets unexpected IR. Please fix.
```suggestion
      auto *GEPCE =
          cast<GetElementPtrConstantExpr>(LI->getPointerOperand());
      IndexValue = cast<Constant>(GEPCE->getOperand(2))
                       ->getUniqueInteger()
                       .getZExtValue();
```